elasticdl/doc/async_sgd_design.md

```diff
@@ -0,0 +1,115 @@
+* Blocked computation: PS has to use a lock for model update with gradients and model read by workers to ensure model consistency.
+
+Asynchronous SGD can avoid the wasted and blocked computation mentioned above with a relaxed model consistency.
+
+* PS will accept all gradients from workers.
+* PS does not locks and supports concurrent model reads and updates.
+
+
+
+
+## Asynchronous SGD
+Let us recall how workers train the model in synchronous SGD. Below is the pseudocode:
+
+```
+for minibatch in training_data:
+    accepted = False
+    while not accepted:
+        local_model，model_version = get_model_from_ps()
+        gradients = compute_gradient(local_model, minibatch)
+        accepted = report_gradient_to_ps(gradients, model_version)  
+```
+
+In asynchronous SGD, each worker is training the model in nearly the same way as synchronous SGD. The only difference is that the worker does not need to retrain any minibatch data as PS accepts all gradients. 
+
+```
+for minibatch in training_data:
+    local_model, model_version = get_model_from_ps()
+    gradients = compute_gradient(local_model, minibatch)
+    report_gradient_to_ps(gradients, model_version)  
+```
+
+PS does not need locks in `GetModel` and `ReportGradient` GRPC services for asynchronous SGD. 
+
+```
+def GetModel():
+    pb_model = Model()
+    for variable in pb_model:
+        assign_value(variable, current_variable_value_in_PS)
+    return pb_model, PS_model_version
+    
+def ReportGradient(gradients, version):
+    grad_var = zip(gradients, model_variables)
+    optimizer.apply_gradient(grad_var)
+    PS_model_version.atomic_add(1)
+```
+
+### Relaxed Model Consistency
+PS can processes multiple GRPC calls `GetModel` and `ReportGradients` concurrently. Thus, there are two kinds of relaxed model consistency.
+
+1. In `GetModel`, during the variable assign loop, there may be `ReportGradient` GRPC service running and updating the variables. Thus, variables in `local_model` in workers may contain values from model versions. `model_version` from `get_model_from_ps` is just a proximate model version.
+2. There may be multiple `ReportGradient` running concurrently. Different model variables may apply these gradients in different orders. 
+
+Also, the concurrent updates to variables in `ReportGradient` may cause some gradients are not applied, as the updates can overwritten by other concurrent running updates. TensorFlow optimizers have an argument [`use_locking`](https://github.com/tensorflow/tensorflow/blob/ff441191277b7e758deb48e45249fee9e880f2c8/tensorflow/python/training/optimizer.py#L319). If [`use_locking`](https://github.com/tensorflow/tensorflow/blob/ff441191277b7e758deb48e45249fee9e880f2c8/tensorflow/python/training/optimizer.py#L319) is `True`, TensorFlow will use a [lock](https://github.com/tensorflow/tensorflow/blob/11e22c01eb801ff24200afcdce8a03a7cdd2ed3f/tensorflow/core/kernels/training_ops.cc#L528) to prevent concurrent updates to variables.
```

Comment:
can be overwritten

Reply:
I think if users set `use_locking` to `True`, the updating process will be serialized. If workers wait for PS's response until PS applies the update, the training speed can be slow.  Assume that workers in the cluster have a similar training speed, it will introduce a lot of conflicts and workers have to wait for a long time until PS applies their update.

Reply:
This lock applies to the individual variable update, not for the whole `optimizer.apply_gradient`. 
For example, if there are two concurrent  `optimizer.apply_gradient(g1)` and `optimizer.apply_gradient(g2)` and there are n variables in the model.  Both `optimizer.apply_gradient(g1)` and `optimizer.apply_gradient(g2)` will aquire n locks in sequence, a lock for update a variable. The performance impact is small as they are likely to aquire different locks at a time.